### PR TITLE
Add prometheus-cpp 0.11.0

### PR DESCRIPTION
**Pull request by Fefer-Ivan**
Specify library name and version:  **prometheus-cpp/0.11.0**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by jgsogo** [comment]
This job and some others in the last hours have failed with a _"Unable to retrieve commit message"_ on our CI. I cannot tell if it is been a glitch in the CI or in the communication with Github. We will try to catch this error and trigger the build automatically.... meanwhile, I've triggered it manually.

Thanks for your contribution!

File: recipes/prometheus-cpp/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        self.cpp_info.components["core"].names["cmake_find_package_multi"] = "core"
+        self.cpp_info.components["core"].names["pkg_config"] = "core"
+        self.cpp_info.components["core"].libs = ["prometheus-cpp-core"]
+
+        if self.options.with_push:
+            self.cpp_info.components["push"].names["cmake_find_package"] = "push"
+            self.cpp_info.components["push"].names["cmake_find_package_multi"] = "push"
+            self.cpp_info.components["push"].names["pkg_config"] = "push"
+            self.cpp_info.components["push"].libs = ["prometheus-cpp-push"]
+            self.cpp_info.components["push"].requires = ["core", "libcurl::libcurl"]
```
Comment (SpaceIm) [inline recipes/prometheus-cpp/all/conanfile.py:102]:
```suggestion
            self.cpp_info.components["push"].requires = ["core", "libcurl::libcurl"]
            if self.settings.os == "Linux":
                self.cpp_info.components["push"].system_libs = ["pthread", "rt"]
```

https://github.com/jupp0r/prometheus-cpp/blob/8a7450115857b1d27ecae89636be0938f93bc996/push/CMakeLists.txt#L14

File: recipes/prometheus-cpp/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+    def package_info(self):
+        self.cpp_info.names["cmake_find_package"] = "prometheus-cpp"
+        self.cpp_info.names["cmake_find_package_multi"] = "prometheus-cpp"
+        self.cpp_info.names["pkg_config"] = "prometheus-cpp"
+
+        self.cpp_info.components["core"].names["cmake_find_package"] = "core"
+        self.cpp_info.components["core"].names["cmake_find_package_multi"] = "core"
+        self.cpp_info.components["core"].names["pkg_config"] = "core"
+        self.cpp_info.components["core"].libs = ["prometheus-cpp-core"]
```
Comment (SpaceIm) [inline recipes/prometheus-cpp/all/conanfile.py:95]:
```suggestion
        self.cpp_info.components["core"].libs = ["prometheus-cpp-core"]
        if self.settings.os == "Linux":
            self.cpp_info.components["core"].system_libs = ["pthread", "rt"]
```

https://github.com/jupp0r/prometheus-cpp/blob/8a7450115857b1d27ecae89636be0938f93bc996/core/CMakeLists.txt#L22

File: recipes/prometheus-cpp/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    def package(self):
+        self.copy(os.path.join(self._source_subfolder, "LICENSE"), dst="licenses")
+        cmake = self._configure_cmake()
+        cmake.install()
+        tools.rmdir(os.path.join(self.package_folder, "lib", "cmake"))
+
+    def package_info(self):
+        self.cpp_info.names["cmake_find_package"] = "prometheus-cpp"
+        self.cpp_info.names["cmake_find_package_multi"] = "prometheus-cpp"
+        self.cpp_info.names["pkg_config"] = "prometheus-cpp"
```
Comment (SpaceIm) [inline recipes/prometheus-cpp/all/conanfile.py:90]:
```suggestion
```

There is no official pkg-config file. This value will be the default for pkg-config generator anyway, but it's misleading to enforce pkg-config name when there is no official pkg-config file.

File: recipes/prometheus-cpp/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+        self.cpp_info.components["core"].names["cmake_find_package"] = "core"
+        self.cpp_info.components["core"].names["cmake_find_package_multi"] = "core"
+        self.cpp_info.components["core"].names["pkg_config"] = "core"
```
Comment (SpaceIm) [inline recipes/prometheus-cpp/all/conanfile.py:94]:
```suggestion
```

File: recipes/prometheus-cpp/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        self.cpp_info.components["core"].libs = ["prometheus-cpp-core"]
+
+        if self.options.with_push:
+            self.cpp_info.components["push"].names["cmake_find_package"] = "push"
+            self.cpp_info.components["push"].names["cmake_find_package_multi"] = "push"
+            self.cpp_info.components["push"].names["pkg_config"] = "push"
```
Comment (SpaceIm) [inline recipes/prometheus-cpp/all/conanfile.py:100]:
```suggestion
```

File: recipes/prometheus-cpp/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+            self.cpp_info.components["push"].libs = ["prometheus-cpp-push"]
+            self.cpp_info.components["push"].requires = ["core", "libcurl::libcurl"]
+
+        if self.options.with_pull:
+            self.cpp_info.components["pull"].names["cmake_find_package"] = "pull"
+            self.cpp_info.components["pull"].names["cmake_find_package_multi"] = "pull"
+            self.cpp_info.components["pull"].names["pkg_config"] = "pull"
```
Comment (SpaceIm) [inline recipes/prometheus-cpp/all/conanfile.py:107]:
```suggestion
```

File: recipes/prometheus-cpp/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+            self.cpp_info.components["pull"].libs = ["prometheus-cpp-pull"]
+            self.cpp_info.components["pull"].requires = ["core", "civetweb::civetweb-cpp"]
+            if self.options.with_compression:
+                self.cpp_info.components["pull"].requires.append("zlib::zlib")
```
Comment (SpaceIm) [inline recipes/prometheus-cpp/all/conanfile.py:111]:
```suggestion
                self.cpp_info.components["pull"].requires.append("zlib::zlib")
            if self.settings.os == "Linux":
                self.cpp_info.components["pull"].system_libs = ["pthread", "rt"]
```

https://github.com/jupp0r/prometheus-cpp/blob/8a7450115857b1d27ecae89636be0938f93bc996/pull/CMakeLists.txt#L42

**Review by SpaceIm — COMMENTED** [review]
I would also recommend to qualify component names to avoid conflicts with other recipes in pkg-config generator.

File: recipes/prometheus-cpp/all/conanfile.py (hunk 1 above)
Comment (Fefer-Ivan) [reply recipes/prometheus-cpp/all/conanfile.py]: Threads::Threads is used in core, push and pull, so I added this to all components

**Review by SpaceIm — APPROVED** [review]

**Review by uilianries — APPROVED** [review]
LGTM

File: recipes/prometheus-cpp/all/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+            return self._cmake
+        self._cmake = CMake(self)
+        self._cmake.definitions["USE_THIRDPARTY_LIBRARIES"] = False
+        self._cmake.definitions["ENABLE_TESTING"] = False
+        self._cmake.definitions["OVERRIDE_CXX_STANDARD_FLAGS"] = False
+
+        self._cmake.definitions["BUILD_SHARED_LIBS"] = self.options.shared
+        self._cmake.definitions["ENABLE_PULL"] = self.options.with_pull
+        self._cmake.definitions["ENABLE_PUSH"] = self.options.with_push
+        self._cmake.definitions["ENABLE_COMPRESSION"] = self.options.with_compression
```
Comment (prince-chrismc) [inline recipes/prometheus-cpp/all/conanfile.py:72]: What if libcurl does not have compression is that a problem? https://github.com/conan-io/conan-center-index/blob/8c69557071cf79e94a6807c7b44a101c843da6ea/recipes/libcurl/all/conanfile.py#L152

File: recipes/prometheus-cpp/all/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+            self.cpp_info.components["prometheus-cpp-pull"].names["cmake_find_package_multi"] = "pull"
+            self.cpp_info.components["prometheus-cpp-pull"].libs = ["prometheus-cpp-pull"]
+            self.cpp_info.components["prometheus-cpp-pull"].requires = [
+                "prometheus-cpp-core",
+                "civetweb::civetweb-cpp"
+            ]
+            if self.options.with_compression:
+                self.cpp_info.components["prometheus-cpp-pull"].requires.append("zlib::zlib")
+            if self.settings.os == "Linux":
+                self.cpp_info.components["prometheus-cpp-pull"].system_libs = ["pthread", "rt"]
```
Comment (prince-chrismc) [inline recipes/prometheus-cpp/all/conanfile.py:119]: is this needed? they are specified with the core already

File: recipes/prometheus-cpp/all/conanfile.py [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+            del self.options.fPIC
+        tools.check_min_cppstd(self, 11)
+
+    def requirements(self):
+        if self.options.with_pull:
+            self.requires("civetweb/1.13")
+        if self.options.with_push:
+            self.requires("libcurl/7.74.0")
+        if self.options.with_compression:
+            self.requires("zlib/1.2.11")
```
Comment (prince-chrismc) [inline recipes/prometheus-cpp/all/conanfile.py:54]: Compression only exists on the pull https://github.com/jupp0r/prometheus-cpp/search?q=zlib

This should be conditional and not impact package_id when doing push only

https://github.com/jupp0r/prometheus-cpp/blob/bb017ec15a824d3301845a1274b4b46a01d6d871/pull/CMakeLists.txt#L45

File: recipes/prometheus-cpp/all/conanfile.py (hunk 8 above)
Comment (prince-chrismc) [reply recipes/prometheus-cpp/all/conanfile.py:72]: This does not matter since lib curl is push

**Review by prince-chrismc — CHANGES_REQUESTED** [review]

File: recipes/prometheus-cpp/all/conanfile.py [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+
+class PrometheusCppConan(ConanFile):
+    name = "prometheus-cpp"
+    description = "Prometheus Client Library for Modern C++"
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://github.com/jupp0r/prometheus-cpp"
+    license = "MIT"
+    topics = ("conan", "metrics", "prometheus", "networking")
+    exports_sources = ["CMakeLists.txt"]
+    generators = "cmake", "cmake_find_package", "cmake_find_package_multi"
```
[the remark this answers is not in the document]
Comment (SpaceIm) [reply recipes/prometheus-cpp/all/conanfile.py:13]: Yes but required

File: recipes/prometheus-cpp/all/conanfile.py (hunk 10 above)
Comment (SpaceIm) [reply recipes/prometheus-cpp/all/conanfile.py:54]: Good catch. `self.options.with_compression` should be deleted in `configure()` if `not self.options.with_pull`. Afterwards, call `self.options.get_safe("with_compression")`

I would also remove `with_`prefix to all options.

File: recipes/prometheus-cpp/all/conanfile.py (hunk 9 above)
Comment (SpaceIm) [reply recipes/prometheus-cpp/all/conanfile.py:119]: explicit is better than implicit. They are direct dependencies of these components, all these dependencies are private, so if conan can implement public/private property in the future, this recipe clearly shows dependencies.

**Review by SSE4 — APPROVED** [review]

**Pull request merged**